File: game-server/src/main/java/work/fking/pangya/game/task/HandoverTask.java
package work.fking.pangya.game.task;

import io.netty.channel.Channel;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import work.fking.pangya.game.GameServer;
import work.fking.pangya.game.SessionClient.SessionInfo;
import work.fking.pangya.game.net.ClientPacketDispatcher;
import work.fking.pangya.game.net.ClientPacketType;
import work.fking.pangya.game.net.ClientProtocol;
import work.fking.pangya.game.net.GameProtocolDecoder;
import work.fking.pangya.game.packet.outbound.CookieBalancePacket;
import work.fking.pangya.game.packet.outbound.EquipmentPacket;
import work.fking.pangya.game.packet.outbound.HandoverReplies;
import work.fking.pangya.game.packet.outbound.HandoverReplies.HandoverResult;
import work.fking.pangya.game.packet.outbound.IffContainerChunkPacket;
import work.fking.pangya.game.packet.outbound.MascotRosterPacket;
import work.fking.pangya.game.packet.outbound.PangBalancePacket;
import work.fking.pangya.game.packet.outbound.ServerChannelsPacket;
import work.fking.pangya.game.packet.outbound.TreasureHunterPacket;

public class HandoverTask implements Runnable {

    private static final Logger LOGGER = LoggerFactory.getLogger(HandoverTask.class);
    private static final ClientProtocol PROTOCOL = ClientProtocol.create(ClientPacketType.values());

    private final GameServer gameServer;
    private final Channel channel;
    private final int cryptKey;
    private final String sessionKey;

    public HandoverTask(GameServer gameServer, Channel channel, int cryptKey, String sessionKey) {
        this.gameServer = gameServer;
        this.channel = channel;
        this.cryptKey = cryptKey;
        this.sessionKey = sessionKey;
    }

    @Override
    public void run() {
        SessionInfo sessionInfo;
        try {
            sessionInfo = gameServer.sessionClient().loadSession(sessionKey);
        } catch (Exception e) {
            LOGGER.warn("Handover error sessionKey={}, message={}", sessionKey, e.getMessage());
            channel.writeAndFlush(HandoverReplies.error(HandoverResult.CANNOT_CONNECT_LOGIN_SERVER));
            return;
        }

        if (sessionInfo == null) {
            LOGGER.warn("Handover failed, unknown sessionKey={}", sessionKey);
            channel.writeAndFlush(HandoverReplies.error(HandoverResult.CANNOT_CONNECT_LOGIN_SERVER));
            return;
        }
        channel.write(HandoverReplies.ok());

        // load all the player stuff
        for (int i = 1; i <= 15; i++) {
            try {
                // do some fancy fake loading
                channel.writeAndFlush(HandoverReplies.updateProgressBar(i));
                Thread.sleep(25);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
        // if no error, register it
        var player = gameServer.registerPlayer(channel, sessionInfo.uid(), sessionInfo.nickname());

        // modify the pipeline
        var pipeline = channel.pipeline();
        pipeline.remove("handoverHandler");
        pipeline.addLast("decoder", GameProtocolDecoder.create(PROTOCOL, cryptKey));
        pipeline.addLast("packetDispatcher", ClientPacketDispatcher.create(gameServer, player, PROTOCOL.handlers()));

        channel.write(HandoverReplies.handoverReply(player));
        IffContainerChunkPacket.create(0x70, player.characterRoster()).forEach(channel::write);
        IffContainerChunkPacket.create(0x71, player.caddieRoster()).forEach(channel::write);
        channel.write(EquipmentPacket.create(player));
        IffContainerChunkPacket.create(0x73, player.inventory()).forEach(channel::write);
        channel.write(new TreasureHunterPacket());
        channel.write(new MascotRosterPacket());
        channel.write(CookieBalancePacket.create(player));
        channel.write(PangBalancePacket.create(player));
        channel.writeAndFlush(new ServerChannelsPacket());
    }
}

File: game-server/src/main/kotlin/work/fking/pangya/game/task/HandoverTask.kt
package work.fking.pangya.game.task

import io.netty.channel.Channel
import org.slf4j.LoggerFactory
import work.fking.pangya.game.GameServer
import work.fking.pangya.game.session.SessionClient
import work.fking.pangya.game.net.ClientPacketDispatcher
import work.fking.pangya.game.net.ClientPacketType
import work.fking.pangya.game.net.ClientProtocol
import work.fking.pangya.game.net.ProtocolDecoder
import work.fking.pangya.game.packet.outbound.CardholicReplies
import work.fking.pangya.game.packet.outbound.CookieBalancePacket
import work.fking.pangya.game.packet.outbound.EquipmentPacket
import work.fking.pangya.game.packet.outbound.HandoverReplies
import work.fking.pangya.game.packet.outbound.HandoverReplies.HandoverResult
import work.fking.pangya.game.packet.outbound.MascotRosterPacket
import work.fking.pangya.game.packet.outbound.PangBalancePacket
import work.fking.pangya.game.packet.outbound.ServerChannelsPacket
import work.fking.pangya.game.packet.outbound.TreasureHunterPacket
import work.fking.pangya.game.packet.outbound.chunkIffContainer
import work.fking.pangya.game.session.SessionInfo
import java.util.concurrent.TimeUnit

private val LOGGER = LoggerFactory.getLogger(HandoverTask::class.java)
private val PROTOCOL = ClientProtocol(ClientPacketType.entries.toTypedArray())

class HandoverTask(
    private val gameServer: GameServer,
    private val channel: Channel,
    private val cryptKey: Int,
    private val loginKey: String
) : Runnable {

    override fun run() {
        val sessionInfo: SessionInfo? = try {
            gameServer.sessionClient.loadSessionInfo(loginKey)
        } catch (e: Exception) {
            LOGGER.warn("Handover error loginKey={}, message={}", loginKey, e.message)
            channel.writeAndFlush(HandoverReplies.error(HandoverResult.CANNOT_CONNECT_LOGIN_SERVER))
            return
        }
        if (sessionInfo == null) {
            LOGGER.warn("Handover failed, unknown loginKey={}", loginKey)
            channel.writeAndFlush(HandoverReplies.error(HandoverResult.CANNOT_CONNECT_LOGIN_SERVER))
            return
        }
        channel.write(HandoverReplies.ok())

        val playerUid = sessionInfo.uid

        // The player picked a new character on the login server, this is a brand-new account
        sessionInfo.characterIffId?.let {
            val future = gameServer.submitTask(
                NewPlayerSetupTask(
                    playerUid = sessionInfo.uid,
                    characterIffId = sessionInfo.characterIffId,
                    characterHairColor = sessionInfo.characterHairColor ?: 0,
                    persistenceCtx = gameServer.persistenceCtx
                )
            )
            try {
                future.get(10, TimeUnit.SECONDS)
            } catch (e: Exception) {
                LOGGER.warn("Failed to process playerId $playerUid handover", e)
                channel.disconnect()
                return
            }
        }

        val persistenceCtx = gameServer.persistenceCtx

        val updateNicknameFuture = gameServer.submitTask { persistenceCtx.noTx { tx -> playerRepository.updateNickname(tx, playerUid, sessionInfo.nickname) } }
        val playerWalletFuture = gameServer.submitTask { persistenceCtx.noTx { tx -> playerRepository.loadWallet(tx, playerUid) } }
        val characterRosterFuture = gameServer.submitTask { persistenceCtx.noTx { tx -> characterRepository.loadRoster(tx, playerUid) } }
        val caddieRosterFuture = gameServer.submitTask { persistenceCtx.noTx { tx -> caddieRepository.loadRoster(tx, playerUid) } }
        val inventoryFuture = gameServer.submitTask { persistenceCtx.noTx { tx -> inventoryRepository.load(tx, playerUid) } }
        val cardInventoryFuture = gameServer.submitTask { persistenceCtx.noTx { tx -> cardRepository.load(tx, playerUid) } }
        val equipmentFuture = gameServer.submitTask { persistenceCtx.noTx { tx -> equipmentRepository.load(tx, playerUid) } }
        val statisticsFuture = gameServer.submitTask { persistenceCtx.noTx { tx -> statisticsRepository.load(tx, playerUid) } }
        val achievementsFuture = gameServer.submitTask { persistenceCtx.noTx { tx -> achievementsRepository.load(tx, playerUid) } }

        val futures = listOf(
            updateNicknameFuture,
            playerWalletFuture,
            characterRosterFuture,
            caddieRosterFuture,
            inventoryFuture,
            cardInventoryFuture,
            equipmentFuture,
            statisticsFuture,
            achievementsFuture
        )

        var progress = 1
        for (future in futures) {
            try {
                future.get(10, TimeUnit.SECONDS)
            } catch (e: Exception) {
                LOGGER.warn("Failed to process playerId $playerUid handover", e)
                channel.disconnect()
                futures.forEach { it.cancel(true) }
                return
            }
            channel.writeAndFlush(HandoverReplies.updateProgressBar(progress))
            progress++
        }

        val player = gameServer.registerPlayer(
            channel = channel,
            sessionInfo = sessionInfo,
            playerWallet = playerWalletFuture.get(),
            characterRoster = characterRosterFuture.get(),
            caddieRoster = caddieRosterFuture.get(),
            inventory = inventoryFuture.get(),
            cardInventory = cardInventoryFuture.get(),
            equipment = equipmentFuture.get(),
            statistics = statisticsFuture.get(),
            achievements = achievementsFuture.get()
        )

        // modify the pipeline
        val pipeline = channel.pipeline()
        pipeline.remove("handoverHandler")
        pipeline.remove("timeoutHandler")
        pipeline.addLast("decoder", ProtocolDecoder(PROTOCOL, cryptKey))
        pipeline.addLast("packetDispatcher", ClientPacketDispatcher(gameServer, player, PROTOCOL.handlers()))


        channel.write(HandoverReplies.handoverReply(player))
        chunkIffContainer(0x70, player.characterRoster).forEach(channel::write)
        chunkIffContainer(0x71, player.caddieRoster).forEach(channel::write)
        channel.write(EquipmentPacket(player))
        chunkIffContainer(0x73, player.inventory).forEach(channel::write)
        channel.write(TreasureHunterPacket())
        channel.write(MascotRosterPacket())
        channel.write(CookieBalancePacket(player))
        channel.write(PangBalancePacket(player))
        channel.write(CardholicReplies.inventory(player.cardInventory))
        channel.writeAndFlush(ServerChannelsPacket(gameServer.serverChannels))
    }
}
